packages/engine/lib/error/examples/hooks.rs

```diff
@@ -0,0 +1,74 @@
+use std::{
+    collections::hash_map::{Entry, HashMap},
+    fmt,
+};
+
+use error::{bail, set_display_hook, Result};
+use provider::{Demand, Provider};
+
+#[derive(Debug)]
+enum ErrorKind {
+    OccupiedEntry(&'static str, u64),
+}
+
+impl Provider for ErrorKind {
+    fn provide<'a>(&'a self, demand: &mut Demand<'a>) {
+        match self {
+            ErrorKind::OccupiedEntry(key, value) => {
+                demand.provide_ref(*key).provide_ref(value);
+            }
+        }
+    }
+}
+
+impl fmt::Display for ErrorKind {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            ErrorKind::OccupiedEntry(key, value) => {
+                write!(fmt, "Entry {key:?} is already occupied by {value}")
+            }
+        }
+    }
+}
+
+fn create_new_entry(
+    map: &mut HashMap<&str, u64>,
+    key: &'static str,
+    value: u64,
+) -> Result<(), ErrorKind> {
+    match map.entry(key) {
+        Entry::Occupied(entry) => {
+            // `bail!` returns `Err(Report)` constructed from it's parameters
+            bail!(context: ErrorKind::OccupiedEntry(key, *entry.get()))
+        }
+        Entry::Vacant(entry) => {
+            entry.insert(value);
+        }
+    }
+    Ok(())
+}
+
+fn main() -> Result<(), impl fmt::Debug> {
+    // When calling the `Display` implementation, this hook will be executed instead of the
+    // implementation above
+    set_display_hook(|report, fmt| {
+        if let Some(key) = report.request_ref::<str>().next() {
+            write!(fmt, "'{key}' is occupied")
```

Comment:
I'm also confused as to why this is specific to an ErrorKind, as this display hook is globally for any `Report` right? If we had another error within the Report that had a `str` context, then we're going to have something misleading where it's 
> 'some random string' is occupied

I would see this being more useful if we're trying to have a generalised implementation that calls debug and display respectively on the associated contexts. I'm not sure I like the fact that the API has to pre-emptively know about `str` etc. though, rather than these somehow being iterable.

Reply:
I think this is meant to be something like a logging formatter where you have pretty format or a concise, etc.

Maybe we could give an example of a JSON one?